### PR TITLE
[FIX] stock_mts_mto_rule: fix routes to supply

### PR DESCRIPTION
**Pull request by umiphos**
### This MR is an effort to fix a bug
To replicate the bug, follow this steps:

1.  Create a product that it's routes are **Manufacture** and **Make to Order**(found in Inventory tab)
2. Add materials to that product:
 a. To material A, add a vendor, add **Buy** and **Make to Order**
 b. To material B, add a vendor, add **Buy** and **Make To Order + Make To Stock**
3. Create a SO with that product(the first one)

To check the bug, go to **Purchase** and you will see that one of the materials hasn't had a purchase order. 

### Desired behavior after PR is merged:
That after the test from before, there are 2 PO lines or orders, one for each product material

### Current behavior before PR:
Click on this video to see what it does
[Video about this PR](https://youtu.be/eYoisV5v7RI "Video description")

### CODE REVIEW

File: stock_mts_mto_rule/model/mrp_production.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
```
Comment (rousseldenis) [inline stock_mts_mto_rule/model/mrp_production.py]: Please remove this and put licensing following this format https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py

File: stock_mts_mto_rule/model/mrp_production.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        except UserError:
+            mto_route = False
+        # This new route is for taking in count this module, the rest is the same as odoo # noqa
+        mto_route |= self.env['stock.location.route'].search([
+            ('name', 'like', 'Make To Order + Make To Stock')], limit=1)
+        for move in self.move_raw_ids:
+            product = move.product_id
+            routes = product.route_ids + product.route_from_categ_ids
+            # TODO: optimize with read_group?
+            pull = self.env['procurement.rule'].search([
```
Comment (rousseldenis) [inline stock_mts_mto_rule/model/mrp_production.py]: The better is to do search before the loop and parse result inside.

File: stock_mts_mto_rule/view/warehouse.xml [hunk 3]
```diff
@@ -6,7 +6,7 @@
         <field name="model">stock.warehouse</field>
         <field name="inherit_id" ref="stock.view_warehouse"/>
         <field name="arch" type="xml">
-            <xpath expr="//field[@name='delivery_steps']" position="after">
+            <xpath expr="//field[@name='partner_id']" position="after">
```
Comment (rousseldenis) [inline stock_mts_mto_rule/view/warehouse.xml]: Why this change ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review.

Could you please insert tests for your method change ?

File: stock_mts_mto_rule/model/mrp_production.py (hunk 1 above)
Comment (rousseldenis) [reply stock_mts_mto_rule/model/mrp_production.py]: And you can remove utf-8 header

File: stock_mts_mto_rule/view/warehouse.xml (hunk 3 above)
Comment (umiphos) [reply stock_mts_mto_rule/view/warehouse.xml]: if you install this all by itself, the checkbox for the field is not displayed, that's why I saw this modification as needed

**Comment by umiphos** [comment]
@moylop260 can you give me your opinion about this change?
Context for the solution:
We need that this module takes the buy route if we select this mdule MTO+MTS, odoo, by default only takes MTS routes

**Comment by moylop260** [comment]
I think that you should use a splitted module for mrp in order to avoid install mrp if you dont likenit

**Comment by rousseldenis** [comment]
> I think that you should use a splitted module for mrp in order to avoid install mrp if you dont likenit

Agree. More coherent

File: stock_mts_mto_mrp/README.rst [hunk 4]
```diff
@@ -0,0 +1,92 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+==================
+Stock MTS+MTO Rule
```
Comment (umiphos) [inline stock_mts_mto_mrp/README.rst]: Please, check the Reame and make improvements on how to use it, if there is a way to add a video, please do it.

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            mto_route = False
+        # This new route is for taking in count this module, the rest is the same as odoo # noqa
+        mto_route |= self.env['stock.location.route'].search([
+            ('name', 'like', 'Make To Order + Make To Stock')], limit=1)
+        for move in self.move_raw_ids:
+            product = move.product_id
+            routes = product.route_ids + product.route_from_categ_ids
+            # TODO: optimize with read_group?
+            pull = self.env['procurement.rule'].search([
+                ('route_id', 'in', [x.id for x in routes]),
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: routes.ids

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                ('location_src_id', '=', move.location_id.id),
+                ('location_id', '=', move.location_dest_id.id)], limit=1)
+            if pull and (pull.procure_method == 'make_to_order'):
+                move.procure_method = pull.procure_method
+            elif not pull:  # If there is no make_to_stock rule either
+                intersection = set(mto_route.ids).intersection(
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: `mto_route & routes`

File: stock_mts_mto_mrp/oca_dependencies.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+addons-vauxoo git@github.com:OCA/stock-logistics-warehouse.git
```
Comment (moylop260) [inline stock_mts_mto_mrp/oca_dependencies.txt]: Use `stock-logistics-warehouse`

File: stock_mts_mto_rule/__manifest__.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
  'category': 'Warehouse',
  'summary': 'Add a MTS+MTO route',
  'depends': ['stock',
+             'mrp',
```
Comment (moylop260) [inline stock_mts_mto_rule/__manifest__.py]: What is the focus to split the modules if you still need mrp here

File: stock_mts_mto_rule/tests/test_mto_mts_route.py [hunk 9]
```diff
@@ -142,6 +142,38 @@ def test_rename_warehouse(self):
         self.warehouse.name = new_warehouse_name
         self.assertEqual(new_rule_name, self.warehouse.mts_mto_rule_id.name)
 
+    def test_standard_mto_mts_buy_route(self):
```
Comment (moylop260) [inline stock_mts_mto_rule/tests/test_mto_mts_route.py]: Shouldn't be a tests for mrp?

File: stock_mts_mto_mrp/README.rst [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/stock-logistics-warehouse/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed feedback.
+
+
+Contributors
+------------
+
+* Arturo Flores <umiphos@gmail.com>
```
Comment (moylop260) [inline stock_mts_mto_mrp/README.rst]: vx email

File: stock_mts_mto_rule/tests/test_mto_mts_route.py (hunk 9 above)
Comment (umiphos) [reply stock_mts_mto_rule/tests/test_mto_mts_route.py]: yes, but also I added the buy route, that's an important part from the base mto+mts, that it does not make the buy from the vendor

File: stock_mts_mto_rule/__manifest__.py (hunk 8 above)
Comment (umiphos) [reply stock_mts_mto_rule/__manifest__.py]: my :bat:

File: stock_mts_mto_mrp/model/mrp_production.py (hunk 6 above)
Comment (umiphos) [reply stock_mts_mto_mrp/model/mrp_production.py]: I've so much to learn....

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            # TODO: optimize with read_group?
+            pull = self.env['procurement.rule'].search([
+                ('route_id', 'in', [x for x in routes.ids]),
+                ('location_src_id', '=', move.location_id.id),
+                ('location_id', '=', move.location_dest_id.id)], limit=1)
+            if pull and (pull.procure_method == 'make_to_order'):
+                move.procure_method = pull.procure_method
+            elif not pull:  # If there is no make_to_stock rule either
+                intersection = mto_route & routes
+                if mto_route and intersection:
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: Is not redundant validate `mto_route and intersection` considering that intersection has mto_route too?
In fact, you can re-use the `elif not pull` for `elif not pull and mto_route & routes`

I mean,

```python
            if pull and (pull.procure_method == 'make_to_order'):
                move.procure_method = pull.procure_method
            elif not pull and mto_route & routes:  # If there is no make_to_stock rule either
                move.procure_method = 'make_to_order'
```

File: stock_mts_mto_mrp/oca_dependencies.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+stock-logistics-warehouse git@github.com:OCA/stock-logistics-warehouse.git
```
Comment (moylop260) [inline stock_mts_mto_mrp/oca_dependencies.txt]: Use just `stock-logistics-warehouse`
The URL to OCA is auto-completed

File: stock_mts_mto_rule/tests/test_mto_mts_route.py [hunk 13]
```diff
@@ -142,6 +142,38 @@ def test_rename_warehouse(self):
...
+        mto_route = self.env.ref('stock.route_warehouse0_mto')
+        manufacture_route = self.env['stock.location.route'].search([
+            ('name', 'ilike', 'Manufacture')])
+        vendor = self.env['res.partner'].search([], limit=1)
+
+        product.write({'seller_ids': [(6, 0, [vendor.id])]})
+        product.route_ids = [(6, 0, [mto_route.id, manufacture_route.id])]
+
+        bom_bom = self.env.ref('product.product_product_3')
+        bom_material = self.env['mrp.bom'].create({
```
Comment (moylop260) [inline stock_mts_mto_rule/tests/test_mto_mts_route.py]: The module `stock_mts_mto_rule` doesn't depend of `mrp` anymore

If you run the tests of this module without install another one you will raise an error because mrp.bom model is not defined.

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    def _adjust_procure_method(self):
+        """This method is taken from odoo as is from the module mrp. This is
+        done because there was no way to interfiere in the decision making
+        about what routes to take"""
+        try:
+            mto_route = self.env['stock.warehouse']._get_mto_route()
+        except UserError:
+            mto_route = False
+        # This new route is for taking in count this module, the rest is the same as odoo # noqa
+        mto_route |= self.env['stock.location.route'].search([
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: If there is a UserError exception then you will raise a error here of 
`False |= browse`

Better use:
```python
        try:
            mto_route = self.env['stock.warehouse']._get_mto_route()
        except UserError:
            mto_route = self.env['stock.warehouse']
```

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+            ('name', 'like', 'Make To Order + Make To Stock')], limit=1)
+        for move in self.move_raw_ids:
+            product = move.product_id
+            routes = product.route_ids + product.route_from_categ_ids
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: | instead of + since that could be duplicated

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+            # TODO: optimize with read_group?
+            pull = self.env['procurement.rule'].search([
+                ('route_id', 'in', [x for x in routes.ids]),
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: ('route_id', 'in', routes.ids)

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+                ('location_src_id', '=', move.location_id.id),
+                ('location_id', '=', move.location_dest_id.id)], limit=1)
+            if pull and (pull.procure_method == 'make_to_order'):
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: If pull is a BrowseNull then you will have `pull.procure_method` with result Null
Then you don't need validate 2 times `if pull and pull.`

You can use just: `if pull.procure_method == 'make_to_order'`

File: stock_mts_mto_mrp/model/mrp_production.py (hunk 14 above)
Comment (umiphos) [reply stock_mts_mto_mrp/model/mrp_production.py]: thanks, should I propose(later in my free time) a fix to the odoo too?

File: stock_mts_mto_rule/tests/test_mto_mts_route.py (hunk 13 above)
Comment (umiphos) [reply stock_mts_mto_rule/tests/test_mto_mts_route.py]: removed everything I was editing from mts+mto module

**Review by moylop260 — APPROVED** [review]

**Comment by moylop260** [comment]
@rousseldenis 
Could you review it, please?

File: stock_mts_mto_mrp/__init__.py [hunk 18]
```diff
@@ -0,0 +1,3 @@
+# Copyright 2018 Vauxoo, www.vauxoo.com
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from . import model
```
Comment (rousseldenis) [inline stock_mts_mto_mrp/__init__.py]: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#14directories

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Minor comment about directory name.

Even if the inside loop search inside the loop is limited to one record, finding a solution to retrieve all records at once is better. Have you try with huge records in DB (move_raw_ids)?

File: stock_mts_mto_mrp/model/mrp_production.py [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    def _adjust_procure_method(self):
+        """This method is taken from odoo as is from the module mrp. This is
+        done because there was no way to interfiere in the decision making
+        about what routes to take"""
+        try:
+            mto_route = self.env['stock.warehouse']._get_mto_route()
+        except UserError:
+            mto_route = self.env['stock.location.route']
+        # This new route is for taking in count this module, the rest is the same as odoo # noqa
+        mto_route |= self.env['stock.location.route'].search([
```
Comment (moylop260) [inline stock_mts_mto_mrp/model/mrp_production.py]: The unique part that you are adding for this custom method is concatenating a new mto_route so you should chagne the method `_get_mto_route` instead.

I mean,

```python
...
def _get_mto_route(self):
	mto_route = self.env['stock.location.route'].search(
		[('name', 'like', 'Make To Order + Make To Stock')], limit=1)
	return mto_route or super()._get_mto_route()
```

Comment (moylop260) [reply stock_mts_mto_mrp/model/mrp_production.py]: another important point, use the translation like the original method:
https://github.com/odoo/odoo/blob/717b7249d5d47421894052e8ad8dc67bc6f3603f/addons/stock/models/stock_warehouse.py#L479

Comment (umiphos) [reply stock_mts_mto_mrp/model/mrp_production.py]: I will do(try to make) the fix so adding another another route similar to mto will work, I think that your reviews will make an point in that PR... but I will not make an enhance to their efficiency...

Comment (umiphos) [reply stock_mts_mto_mrp/model/mrp_production.py]: I will do(try to make) the fix so adding another route similar to mto will work, I think that your reviews will make a good point in that PR... but I will not make an enhance to their efficiency(at the moment)...

Comment (umiphos) [reply stock_mts_mto_mrp/model/mrp_production.py]: Just an update, didn't make that PR because we have deeper problems with the logic of MTS+MTO, odoo is not fit(yet) to store this method, so, after adding this fix to one of our modules we found that we got another problem:

The materials with MTS+MTO in a product aren't going to ever touch our module, so, the decision between stock and purchase is never triggered. If we added the logic that this action is triggered in MRP we enter into a loop...right now I'm out of ideas and the team is looking forward a solution.

Comment (saxomanu) [reply stock_mts_mto_mrp/model/mrp_production.py]: Hi, What's the difference between this PR and this : https://github.com/OCA/manufacture/tree/11.0/mrp_mto_with_stock/
In my mind, with mrp_mto_with_stock the job is done, but you need to clic on "check avaibility" on a MO, and it's very hard when you have a lot of MO.
Why don't you improve mrp_mto_with_stock ?

**Comment by saxomanu** [comment]
Hi, What's the difference between this PR and this : https://github.com/OCA/manufacture/tree/11.0/mrp_mto_with_stock/
In my mind, with mrp_mto_with_stock the job is done, but you need to clic on "check avaibility" on a MO, and it's very hard when you have a lot of MO.
Why don't you improve mrp_mto_with_stock ?

**Comment by bodedra** [comment]
@saxomanu IMO these module https://github.com/OCA/manufacture/tree/11.0/mrp_auto_assign will reserve the raw material moves automatically at creation. If you want, you may improve that module and submit PR.

**Comment by florian-dacosta** [comment]
IMO, there is problem with module stock_mts_mto_rule and mrp_mto_with_stock.
As they are 2 separated modules trying to fix the same problem, the first one for bought products or top level manufactured product and the second for raw materials of a MO...
I guess the module mrp_mto_with_stock should be rethinked, to depends on the module stock_mts_mto_rule. 
Because today, people are kind of lost and do not understand which modules install and why and they have to make 2 configuration on 1 product for the same thing.

It is more or less what you do in this PR by making another module, but
- I think it should be auto-installed
- I think it should be proposed only in version 12, as a replacement of module mrp_mto_with_stock from manufacture REPO, else it will be even worse to have one more conccurent module...
- And mainly, the present implementation of this module do not solve the problem of 'mts+mto' for raw material, as it always force make to order, which is not what a user expects at all !
So today, using stock_mto_mts_rule + mrp_mto_with_stock better answer the need even if it is more configuration.

Actually without this module raw material products configured with mts+mto rule behave like mts products. 
With this module, they behave like mto products.
But neither situation is the one we want!

Maybe this module could be kind of merged with mrp_mto_with_stock in version 12... So it would really do the job, and people would not be lost with 2 different systems to attend the same need.
But until this module does the job (really make mts+mto for raw material), it should not be here...

That said, I think making mrp_mto_with_stock depend on stock_mts_mto_rule is not an easy task and one need to study the feasability seriously...If not possible, then maybe both modules should be rethinked totally to attend to this problem! Sadly, this take a lot of time...

**Comment by umiphos** [comment]
Closing this because we didn't solved all the problems in this version, in future versions we might try to find a better solution for the MTS+MTO
